File: exercism/python/beer-song/beer_song.py
# https://exercism.io/my/solutions/a43b8babd915454c822b5f1bf143ebae


def verse(i):
    VERSE_N = [f'{i} bottles of beer on the wall, {i} bottles of beer.',
               f'Take one down and pass it around, {i - 1} bottle{"s" if i > 2 else ""} of beer on the wall.']
    VERSE_1 = ['1 bottle of beer on the wall, 1 bottle of beer.',
               'Take it down and pass it around, no more bottles of beer on the wall.']
    VERSE_0 = ['No more bottles of beer on the wall, no more bottles of beer.',
               'Go to the store and buy some more, 99 bottles of beer on the wall.']

    if i > 1:
        return VERSE_N
    elif i == 1:
        return VERSE_1
    else:
        return VERSE_0


def recite(start, take=1):
    song = []
    for i in range(start, start - take, -1):
        song.extend(verse(i))
        if i > start - take + 1:
            song.append('')
    return song

File: exercism/elixir/beer-song/lib/beer_song.ex
# https://exercism.io/my/solutions/fc842117968b433daa891a425c1aa398

defmodule BeerSong do
  @spec bottles(integer) :: String.t()
  defp bottles(0), do: "no more bottles"
  defp bottles(1), do: "1 bottle"
  defp bottles(n), do: "#{n} bottles"

  @doc """
  Get a single verse of the beer song
  """
  @spec verse(integer) :: String.t()
  def verse(0) do
    """
    No more bottles of beer on the wall, no more bottles of beer.
    Go to the store and buy some more, 99 bottles of beer on the wall.
    """
  end

  def verse(n) do
    """
    #{bottles(n)} of beer on the wall, #{bottles(n)} of beer.
    Take #{if n === 1, do: "it", else: "one"} down and pass it around, #{bottles(n - 1)} of beer on the wall.
    """
  end

  @spec lyrics() :: String.t()
  def lyrics(), do: lyrics(99..0)

  @doc """
  Get the entire beer song for a given range of numbers of bottles.
  """
  @spec lyrics(Range.t()) :: String.t()
  def lyrics(range) do
    Enum.map_join(range, "\n", &verse(&1))
  end
end
